agent/src/agent.ts

```diff
@@ -481,6 +500,113 @@ export class Agent extends MessageHandler implements ExtensionClient {
             }
         })
 
+        // Store in-memory copy of the most recent Code action
+        const codeActionById = new Map<string, vscode.CodeAction>()
+        this.registerAuthenticatedRequest('codeActions/provide', async (params, token) => {
+            codeActionById.clear()
+            const document = this.workspace.getDocument(vscode.Uri.parse(params.location.uri))
+            if (!document) {
+                throw new Error(`codeActions/provide: document not found for ${params.location.uri}`)
+            }
+            const codeActions: agent_protocol.ProtocolCodeAction[] = []
+            for (const providers of this.codeAction.providers()) {
+                const diagnostics = vscode.languages.getDiagnostics(document.uri)
```

Comment:
nit: can we collect diagnostics once out of the loop?

```suggestion
            const codeActions: agent_protocol.ProtocolCodeAction[] = []
            const diagnostics = vscode.languages.getDiagnostics(document.uri)
            for (const providers of this.codeAction.providers()) {
```

Reply:
Good one! Fixed